### PR TITLE
[3.5] Secure vs insecure image pruning

### PR DESCRIPTION
**Pull request by miminar**
Document new options related to secure connection to integrated docker
registry and a mechanism that decides whether to fall-back to insecure
connection.

Backports #4471

### CODE REVIEW

File: admin_guide/pruning_resources.adoc [hunk 1]
```diff
@@ -190,13 +190,28 @@ particular namespace makes it impossible to calculate their current usage.
...
+.^|`--all`
+|Include images that were not pushed to the registry, but have been mirrored by
+pullthrough. This is on by default. To limit the pruning to images that were
+pushed to the integrated registry, pass `--all=false`.
+
 .^|`--certificate-authority`
 |The path to a certificate authority file to use when communicating with the
 {product-title}-managed registries. Defaults to the certificate authority data
-from the current user's config file.
+from the current user's configuration file. If provided, secure connection will
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: a secure connection

File: admin_guide/pruning_resources.adoc [hunk 2]
```diff
@@ -285,9 +303,41 @@ $ oadm prune images --keep-tag-revisions=3 --keep-younger-than=60m --confirm
 $ oadm prune images --prune-over-size-limit --confirm
 ----
 
+[[pruning-images-secure-or-insecure]]
+=== Using Secure or Insecure Connections
+
+The secure connection is the preferred and recomended approach. It is done over
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: recommended

File: admin_guide/pruning_resources.adoc [hunk 3]
```diff
@@ -339,3 +389,225 @@ Such tags tend to have just one image in its history, which effectively prevents
...
 xref:../dev_guide/managing_images.adoc#tag-naming[Learn more about _istag_
 naming.]
+
+[discrete]
+[[using-secure-connection-against-insecure-registry]]
+==== Using a Secure Connection Against Insecure Registry
+
+If you see a message similar to the following in the output of the `oadm prune
+images`, then your registry is not secured and the `oadm prune images` client
+will attempt to use secure connection:
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: "then your registry is not secured and the `oadm prune images` client
will attempt to use secure connection"

should be 

"then your registry is not secured and the `oadm prune images` client
**attempted** to use secure connection"  ?

File: admin_guide/pruning_resources.adoc [hunk 4]
```diff
@@ -339,3 +389,225 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+
+----
+error: error communicating with registry: Get https://172.30.30.30:5000/healthz: http: server gave HTTP response to HTTPS client
+----
+
+. The recommened solution is to
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: recommended

File: admin_guide/pruning_resources.adoc [hunk 5]
```diff
@@ -339,3 +389,225 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+last component) does not. All unique image blobs become orphans.
+
+- {product-title} refusing an image because of quota restrictions.
+
+- The standard image pruner deleting an image manifest, but is interrupted before
+it deletes the related blobs.
+
+- A bug in the registry pruner, which fails to remove the intended blobs, causing
+the image objects referencing them to be removed and the blobs becoming orphans.
+// Find this BZ
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: we probably shouldn't be documenting bugs (especially ones that are fixed) in our official documentation.

File: admin_guide/pruning_resources.adoc [hunk 6]
```diff
@@ -339,3 +389,225 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+
+To hard prune orphaned blobs from the registry:
+
+. +++<b>Log in:</b>+++ Log in using xref:../cli_reference/get_started_cli.adoc#basic-setup-and-login[the CLI] as a user with an
+xref:../architecture/additional_concepts/authentication.adoc#oauth[access token].
+
+. +++<b>Run a basic image prune:</b>+++ Basic image pruning removes additional
+images that are no longer needed. The hard prune does not remove images on its
+own. It only removes blobs stored in the registry storage. Therefore, you should
+run this just before the hard prune.
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: s/this/a standard prune/

File: admin_guide/pruning_resources.adoc [hunk 7]
```diff
@@ -339,3 +389,225 @@ Such tags tend to have just one image in its history, which effectively prevents
...
+----
+$ oc rollout -n default \
+    latest dc/docker-registry
+----
+
+. +++<b>Add the system:image-pruner role:</b>+++ The service account used to run
+the registry instances requires additional permissions in order to list some
+resources.
+
+.. Get the service account name:
```
Comment (bparees) [inline admin_guide/pruning_resources.adoc]: the registry's service account name

**Review by bparees — CHANGES_REQUESTED** [review]
a few nits, but you'll probably want to wait for @openshift/team-documentation to review as well and then make all the same changes to your 3.4 PR.

**Comment by bparees** [comment]
@openshift/team-documentation ptal

File: admin_guide/pruning_resources.adoc [hunk 8]
```diff
@@ -190,13 +190,28 @@ particular namespace makes it impossible to calculate their current usage.
...
+be initiated.
 
 .^|`--confirm`
-|Indicate that pruning should occur, instead of performing a dry-run.
+|Indicate that pruning should occur, instead of performing a dry-run. This
+requires a valid route to the integrated Docker registry. If this command is
+run outside of the cluster network, the route needs to be provided using
+`--registry-url`.
+
+.^|`--force-insecure`
```
Comment (bmcelvee) [inline admin_guide/pruning_resources.adoc]: We may want to add an additional warning for this option following the table. @bparees what do you think?

Comment (bparees) [reply admin_guide/pruning_resources.adoc]: sure.

File: admin_guide/pruning_resources.adoc [hunk 9]
```diff
@@ -211,13 +226,16 @@ is younger than `<duration>` relative to the current time. (default *60m*)
...
-|===
 
-{product-title} uses the following logic to determine which images and layers to
-prune:
+.^|`--registry-url`
+|The address to use when contacting the registry. The command will attempt to
+use a cluster-internal URL determined from managed images and image streams. In
+case it fails (the registry cannot be resolved or reached), an alternative route
+that works needs to be provided using this flag.
+|===
```
Comment (bmcelvee) [inline admin_guide/pruning_resources.adoc]: s/an alternative route that works needs to be provided using this flag/use this flag to provide an alternate route.

File: admin_guide/pruning_resources.adoc (hunk 3 above)
Comment (miminar) [reply admin_guide/pruning_resources.adoc]: yes

File: admin_guide/pruning_resources.adoc (hunk 8 above)
Comment (miminar) [reply admin_guide/pruning_resources.adoc]: I believe I've covered the danger in the `[pruning-images-secure-or-insecure]` section which is already recommended to read. I would rather refrain from clogging the doc with redundant information.

**Comment by miminar** [comment]
Thank you! Comments addressed. If approved, I'll apply the changes to 3.4 as well.

**Comment by bmcelvee** [comment]
Thanks, @miminar! LGTM. @bparees do you have any additional changes?

**Comment by bparees** [comment]
@bmcelvee nope

**Comment by bmcelvee** [comment]
[rev_history]
|xref:../admin_guide/pruning_resources.adoc#admin-guide-pruning-resources[Pruning Objects]
|Added details on secure versus insecure image pruning, and hard pruning the registry.
%

**Pull request merged**